Added early and late scripts keys and main.sh modification

ColonelPanics opened the pull request:
This feature would allow scripts to be executed either before or after `core` and `scripts. This is particularly useful with local yum repository setup as it allows for the local repo config to be put into place _before_ `base.sh` runs `yum -y update`

ColonelPanics committed:
Added early and late scripts keys and main.sh modification

sierra-tango-echo commented:
LGTM! @BobWhitelock

bobwhitelock commented:
@ColonelPanicks: Couple of questions:

- You've not included the scripts themselves, is that intentional?

- What's the purpose of these groups of scripts - is there a more descriptive name we could give them? Maybe `setup` and `final` or something, if they aren't for a particular purpose, otherwise name them based on the purpose?

- Is it essential to have these as a new namespace of scripts? We already request the `core` scripts all at once ([here](https://github.com/alces-software/metalware-default/blob/a60b08071e3ac211931dc6d5fa200c4a9ce84c71/files/main.sh#L20,L27)) and then run each individually as needed (e.g. [here](https://github.com/alces-software/metalware-default/blob/a60b08071e3ac211931dc6d5fa200c4a9ce84c71/files/main.sh#L40) and [here](https://github.com/alces-software/metalware-default/blob/a60b08071e3ac211931dc6d5fa200c4a9ce84c71/files/core/networking.sh#L23)), as they are order-dependent and we want to run some multiple times. Would it be possible, and maybe make things simpler, to add any new scripts to `core` and then run them using `run_script` in the particular places you need to?

@ste78: this has also prompted me to think that it seems there's scope to improve how we handle the build files, as currently it relies on copying around the code to request, error-check, and run or install groups of files. Maybe we could provide templating macros to perform common templating tasks. Or, possibly better, Metalware/repos should install all files initially itself (since every templated file will need to be requested anyway), and then when templating we can just provide a way to get the paths to the installed files).

sierra-tango-echo commented:
@BobWhitelock those ideas sound good to me, The core idea with these is for us to be able to give a simple instruction along with our external scripts - along the line of "just save the script somewhere on your deployment machine and add its path to the 'prescripts' config array" (ie be able to make modifications to the config only - not have to update the script templates which is more prone to error) That obv needs some sort of skeleton pre-setup in the metalware default repo which is what this PR is our first pass at. Happy for you to rename / rearrange as you see fit, so long as we enable that use case ;)

bobwhitelock committed:
Unify previous examples and need for user scripts

Previously we had `scripts` and `config` namespaces, giving examples of
using the new build files feature. We now want to provide file
namespaces for users to use to run scripts pre and post the `core` setup
scripts.

This commit unifies these two things in this repo, by having `setup` and
`scripts` namespaces to fulfil these two purposes; these now both use
the previous example scripts to demonstrate how these sections could be
used.

bobwhitelock commented:
@ste78: OK, I've noted it as something to look at in the future :).

@ColonelPanicks: I've added a commit to hopefully cover this use case while removing/merging it with the existing example scripts. This repo is kind of functioning as both fuller example and minimal repo for use by users at the moment, so not sure which we want to move towards - if the latter we may want to remove the example scripts entirely. Let me know if this is OK or if it misses something.

ColonelPanics commented on files/main.sh:
```diff
@@ -44,18 +53,4 @@ curl "<%= script.url %>" | /bin/bash
   <% end %>
 <% end %>
 
-mkdir -p /tmp/configs
```
Is the config setup being moved elsewhere or has it been deprecated?

bobwhitelock replied:
The `configs` section was really just an example of how the `files` feature could be used, as the `scripts` section was as well. Do you have existing stuff making use of it at the moment? Happy to add it or something similar back if so.

ColonelPanics replied:
I don't have anything making use of it, that makes sense though. Might be worth leaving in for examples sake though? Don't see it being a huge issue being taken out!

bobwhitelock replied:
It should be fine for now I think, we can always add it back if needed. How this works will probably change and be simplified a bit in the near future anyway, when we look at https://trello.com/c/4xUk3FGp/105-consider-improving-how-nodes-need-to-request-and-interact-with-build-files. I'll merge this through for now :slightly_smiling_face:.

ColonelPanics commented:
Besides the above, happy with the refactor - `setup` seems like a good name for pre-core scripts to run. I think it's worth keeping the example scripts unless that's going to cause any sort of issues leaving them in the scripts namespace.
